apps/vmq_commons/src/gen_mqtt_client.erl

```diff
@@ -775,23 +976,126 @@ maybe_queue_outgoing(PubReq, #state{o_queue = #queue{size = Size, max = Max} = Q
     Size < Max
 ->
     State#state{o_queue = queue_outgoing(PubReq, Q)};
-maybe_queue_outgoing(_PubReq, #state{o_queue = Q} = State) ->
+maybe_queue_outgoing(PubReq, #state{o_queue = Q} = State) ->
     %% drop!
+    lager:debug("Drop from Queue MSG: ~p\n", [trunc_pubreq(PubReq)]),
     State#state{o_queue = drop(Q)}.
 
-queue_outgoing(Msg, #queue{size = Size, queue = QQ} = Q) ->
-    Q#queue{size = Size + 1, queue = queue:in(Msg, QQ)}.
-
-maybe_publish_offline_msgs(#state{o_queue = #queue{size = Size} = Q} = State) when Size > 0 ->
+%% TODO: append to queue
+queue_outgoing(Msg, #queue{queue = QQ} = Q) ->
+    lager:debug("Add to Queue MSG: ~p\n", [trunc_pubreq(Msg)]),
+    NewQQ = replayq:append(QQ, [Msg]),
+    Q#queue{size = replayq:count(NewQQ), queue = NewQQ}.
+
+queue_pubrel(PubRelFrame, QQ, Waiting) ->
+    lager:debug("Add to PubrelQ MSG: ~p\n", [PubRelFrame]),
+    NewQQ = replayq:append(QQ, [PubRelFrame]),
+    {NewQQ, Waiting + 1}.
+
+%% TODO: publish from queue logic
+maybe_publish_offline_msgs(
+    #state{o_queue = #queue{size = Size, out_waiting = Waiting} = Q} = State
+) when Size > 0, Waiting < 1 ->
     publish_from_queue(Q, State);
 maybe_publish_offline_msgs(State) ->
     State.
 
-publish_from_queue(#queue{size = Size, queue = QQ} = Q, State0) when Size > 0 ->
-    {{value, PubReq}, NewQQ} = queue:out(QQ),
-    gen_fsm:send_event(self(), {publish_from_queue, PubReq}),
-    State0#state{o_queue = Q#queue{size = Size - 1, queue = NewQQ}}.
+maybe_publish_pubrel_msgs(
+    #state{pubrel_queue = #queue{size = Size, out_waiting = Waiting} = Q} = State
+) when Size > 0, Waiting < 1 ->
+    publish_from_pubrel_queue(Q, State);
+maybe_publish_pubrel_msgs(
+    #state{pubrel_queue = #queue{size = Size, out_waiting = Waiting} = _} = State
+) when Size > 0, Waiting > 0 ->
+    State;
+maybe_publish_pubrel_msgs(State) ->
+    maybe_publish_offline_msgs(State).
+
+maybe_ack_msgs(_, _, Waiting) when Waiting < 1 ->
+    0;
+maybe_ack_msgs(Queue, AckRef, Waiting) ->
+    ok = replayq:ack(Queue, AckRef),
+    lager:debug("Sent Ack: ~p", [AckRef]),
+    Waiting - 1.
+
+maybe_ack_pubrel(Q, _, Waiting) when Waiting < 1 ->
+    {Q, 0};
+maybe_ack_pubrel(Queue, next, Waiting) ->
+    {NewQ, AckRef, _} = replayq:pop(Queue, #{count_limit => 1}),
+    maybe_ack_pubrel(NewQ, AckRef, Waiting);
+maybe_ack_pubrel(Queue, AckRef, Waiting) ->
+    ok = replayq:ack(Queue, AckRef),
+    lager:debug("Sent Pubrel Ack: ~p", [AckRef]),
+    {Queue, Waiting - 1}.
+
+%% TODO: pop from queue then publish
```

Comment:
todo unclear

Reply:
These TODOs were all old and no longer needed, removed them all in the most recent commit.